builtin/credential/aws/path_role.go

```diff
@@ -41,14 +54,33 @@ that it must match the IAM role ARN specified by this parameter.
 The value is prefix-matched (as though it were a glob ending in
 '*').  The configured IAM user or EC2 instance role must be allowed
 to execute the 'iam:GetInstanceProfile' action if this is
-specified.`,
+specified. This is only checked when auth_type is
+ec2.`,
 			},
 			"bound_iam_instance_profile_arn": {
 				Type: framework.TypeString,
 				Description: `If set, defines a constraint on the EC2 instances to be associated
 with an IAM instance profile ARN which has a prefix that matches
 the value specified by this parameter. The value is prefix-matched
-(as though it were a glob ending in '*').`,
+(as though it were a glob ending in '*'). This is only checked when
+auth_type is ec2.`,
+			},
+			"role_inferred_type": {
+				Type:    framework.TypeString,
+				Default: false,
+				Description: `When auth_type is iam, the
```

Comment:
I'm just not seeing the use case for doing that. If I'm missing something, please inform me!

The way I see it, inferencing basically means, "These additional conditions (i.e., binds) are important to me, so enforce them." Which implies, "If I can't test the conditions, I should just fail" (to be safe). There's simply too much gray area in being "opportunistic" for me to be comfortable with. There are two main ambiguities off the top of my head that I am concerned with:

1. Session ID "looks like" an EC2 instance ID:
  1. Session ID matches an EC2 instance that Vault is able to discover: Probably OK to do inferencing, but requires a request to the AWS APIs to query instance IDs
  1. Session ID doesn't match an EC2 instance that Vault is able to discover. Should we assume alternate roles in different accounts to search for that instance ID? If so, how do we determine the "target" account? If not, do we just ignore the inferencing (and associated bindings)?
1. Session ID doesn't "look like" an EC2 instance ID: probably just ignore bindings completely

Note that Vault would have to keep up to date with what EC2 instance IDs look like.

Anyway, in my mind, the the use of inferencing means "this is important to me and therefore you must apply it and fail if you can't" as opposed to "this seems useful to me and you should apply it if you can and fail open if you can't." If inferencing is requested for a role but can't be used, then I think it's better to fail closed than open.